### PR TITLE
Add tests for https://github.com/openshift/origin/issues/10402

### PR DESCRIPTION
**Pull request by sgallagher**
Proves that generation is always increased when tags are updated.

Signed-off-by: Stephen Gallagher sgallagh@redhat.com

### CODE REVIEW

**Comment by sgallagher** [comment]
@mfojtik This test works, but because there is apparently a settling time for the status contents, I temporarily added a five-second sleep (which is overkill at least on my system). I'm not sure how to get `os::cmd::try_until_success` to deal with something that needs to compare the output with a known value before treating it as success or failure; I've been banging my head against that for the last two hours.

So now I'm just posting what I have and hoping you can help me get across the finish line on this.

**Comment by sgallagher** [comment]
[test]

**Comment by liggitt** [comment]
capture the value you want it to equal, then try_until_text?

File: test/cmd/images_tests.sh [hunk 1]
```diff
@@ -51,6 +51,34 @@ os::cmd::try_until_success 'oc get imagestreamtags wildfly:latest'
...
+# Make sure that switching from a reference tag to a non-reference tag properly
+# updates the generation.
+# We will move the 'latest' tag to point at the same reference as '8.1'.
+orig_spec_tag_generation="$(oc get imagestreams/wildfly --template='{{range .spec.tags}}{{ if eq .name "latest"}}{{ .generation }}{{end}}{{end}}')"
+orig_status_tag_generation="$(oc get imagestreams/wildfly --template='{{range .status.tags}}{{ if eq .tag "latest"}}{{ $item:=index .items 0}}{{$item.generation}}{{end}}{{end}}')"
+reference_81="$(oc get imagestreams/wildfly --template='{{range .status.tags}}{{ if eq .tag "8.1"}}{{ $item:=index .items 0}}{{$item.dockerImageReference}}{{end}}{{end}}')"
+
+# https://github.com/openshift/origin/issues/10402
+# Make latest a --reference to 8.1
+os::cmd::expect_success 'oc tag --source=docker $reference_81 --reference wildfly:latest'
```
Comment (stevekuznetsov) [inline test/cmd/images_tests.sh]:
```sh
os::cmd::expect_success "oc tag --source=docker '${reference_81}' --reference wildfly:latest"
                        ^                       ^ ^            ^^                           ^
```

File: test/cmd/images_tests.sh [hunk 2]
```diff
@@ -51,6 +51,34 @@ os::cmd::try_until_success 'oc get imagestreamtags wildfly:latest'
...
+sleep 5
```
Comment (stevekuznetsov) [inline test/cmd/images_tests.sh]: `sleep`s are fishy -- we have polling functions like `os::cmd::try_until_{text,success,failure}`. Can you use them?

File: test/cmd/images_tests.sh [hunk 3]
```diff
@@ -51,6 +51,34 @@ os::cmd::try_until_success 'oc get imagestreamtags wildfly:latest'
...
+reference_spec_tag_generation="$(oc get imagestreams/wildfly --template='{{range .spec.tags}}{{ if eq .name "latest"}}{{ .generation }}{{end}}{{end}}')"
+reference_status_tag_generation="$(oc get imagestreams/wildfly --template='{{range .status.tags}}{{ if eq .tag "latest"}}{{ $item:=index .items 0}}{{$item.generation}}{{end}}{{end}}')"
+os::cmd::expect_success 'test $reference_spec_tag_generation -gt $orig_spec_tag_generation'
```
Comment (stevekuznetsov) [inline test/cmd/images_tests.sh]:
```sh
os::cmd::expect_success "test '${reference_spec_tag_generation}' -gt '${orig_spec_tag_generation}'"
                        ^     ^ ^                             ^^     ^ ^                        ^^^
```

Comment (stevekuznetsov) [inline test/cmd/images_tests.sh]: These fail correctly if either generation isn't an integer, right?

File: test/cmd/images_tests.sh (hunk 2 above)
Comment (sgallagher) [inline test/cmd/images_tests.sh]: See the comment above; I am trying to find a way to use the `os::cmd::try_until_{text,success,failure}` routines, but I was having trouble and hence asking for advice in this ticket.

I think I had another idea this morning, so I'm going to try that and see if it works.

File: test/cmd/images_tests.sh (hunk 3 above)
Comment (sgallagher) [inline test/cmd/images_tests.sh]: Yes, "test" will just return error code 2 if it gets a non-integer value.

**Comment by sgallagher** [comment]
@stevekuznetsov @liggitt OK, I think I figured out how to avoid those sleeps (and I fixed the minor issues Steve identified as well). PTAL

File: test/cmd/images_tests.sh [hunk 4]
```diff
@@ -51,6 +51,31 @@ os::cmd::try_until_success 'oc get imagestreamtags wildfly:latest'
...
+# updates the generation.
+# We will move the 'latest' tag to point at the same reference as '8.1'.
+orig_spec_tag_generation="$(oc get imagestreams/wildfly --template='{{range .spec.tags}}{{ if eq .name "latest"}}{{ .generation }}{{end}}{{end}}')"
+reference_81="$(oc get imagestreams/wildfly --template='{{range .status.tags}}{{ if eq .tag "8.1"}}{{ $item:=index .items 0}}{{$item.dockerImageReference}}{{end}}{{end}}')"
+
+# https://github.com/openshift/origin/issues/10402
+# Make latest a --reference to 8.1
+os::cmd::expect_success "oc tag --source=docker '${reference_81}' --reference wildfly:latest"
+reference_spec_tag_generation="$(oc get imagestreams/wildfly --template='{{range .spec.tags}}{{ if eq .name "latest"}}{{ .generation }}{{end}}{{end}}')"
+os::cmd::expect_success 'test $reference_spec_tag_generation -gt $orig_spec_tag_generation'
```
Comment (stevekuznetsov) [inline test/cmd/images_tests.sh]: Previous comment seems to have gotten lost: https://github.com/openshift/origin/pull/10622#discussion_r76258407

File: test/cmd/images_tests.sh [hunk 5]
```diff
@@ -66,9 +91,9 @@ os::cmd::expect_success "oc annotate --context='${cluster_admin_context}' --over
 os::cmd::expect_failure_and_text "oc run vulnerable --image=${repository}:new --restart=Never" 'spec.containers\[0\].image: Forbidden: this image is prohibited by policy'
 
 # test image stream tag operations
-os::cmd::expect_success_and_text 'oc get istag/wildfly:latest -o jsonpath={.generation}' '2'
```
Comment (stevekuznetsov) [inline test/cmd/images_tests.sh]: Are these test still valid with your changes, given you're having to change what they're testing for? Also, are they doing something worthwhile? Why is `5` an important generation? It seems a little weird that we're testing `oc get ist` without *doing* something to them first.

**Pull request closed without merging**